x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts

```diff
@@ -911,17 +929,44 @@ export class ObservabilityAIAssistantClient {
     return this.dependencies.knowledgeBaseService.deleteEntry({ id });
   };
 
-  private addResponseLanguage = (messages: Message[], responseLanguage: string): Message[] => {
-    const [systemMessage, ...rest] = messages;
+  private resolveInstructions = async (
+    requestInstructions: Array<string | UserInstruction>,
+    user: { name: string },
+    namespace: string
+  ) => {
+    const knowledgeBaseInstructions = await this.dependencies.knowledgeBaseService.getInstructions(
+      user,
+      namespace
+    );
 
-    const extendedSystemMessage: Message = {
-      ...systemMessage,
-      message: {
-        ...systemMessage.message,
-        content: `You MUST respond in the users preferred language which is: ${responseLanguage}. ${systemMessage.message.content}`,
-      },
-    };
+    if (requestInstructions.length + knowledgeBaseInstructions.length === 0) {
+      return '';
+    }
+
+    const overrides = requestInstructions.filter(
+      (instruction): instruction is UserInstruction => typeof instruction !== 'string'
+    );
+    const overrideIds = overrides.map((instruction) => instruction.doc_id);
+
+    const plainInstructions = requestInstructions
+      .filter((instruction): instruction is string => typeof instruction === 'string')
+      .map<UserInstruction>((instruction, index) => ({
+        doc_id: `${index}-${instruction.substring(0, 10)}`, // This will be thrown away further down
+        text: instruction,
+      }));
+
+    const kept = knowledgeBaseInstructions.filter(
+      (instruction) => !overrideIds.includes(instruction.doc_id)
+    );
+
+    const instructions = [...overrides, ...plainInstructions, ...kept].map(
+      (instruction) => instruction.text
+    );
```

Comment:
what about something like:

```ts
const plainInstructions = requestInstructions.map(
  instruction => typeof instruction === 'string' ? { doc_id: uuid.v4(), text: instruction } : instruction
);
const overrideIds = plainInstructions.map(instruction => instruction.doc_id);
const instructions = plainInstructions.concat(kept.filter(instruction => !overrideIds.contains(instruction.doc_id)));
```

to save some lines and type assertions?

Reply:
Neat! 
Fixed in https://github.com/elastic/kibana/pull/180263/commits/b8b677b37126cf2abd06924612cf7b2e6486b2f6